core/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/ruin/ListRuinRecreateMove.java

```diff
@@ -84,11 +81,23 @@ protected void doMoveOnGenuineVariables(ScoreDirector<Solution_> scoreDirector)
         constructionHeuristicPhase.solvingEnded(solverScope);
         scoreDirector.triggerVariableListeners();
 
+        var entityToInsertedValuesMap = CollectionUtils.<Object, List<Object>> newIdentityHashMap(0);
+        for (var entity : entityToOriginalPositionMap.keySet()) {
+            entityToInsertedValuesMap.put(entity, new ArrayList<>());
+        }
+
         for (var ruinedValue : ruinedValueList) {
-            var location = (LocationInList) listVariableStateSupply.getLocationInList(ruinedValue);
+            var location = listVariableStateSupply.getLocationInList(ruinedValue)
+                    .ensureAssigned();
             entityToNewPositionMap.computeIfAbsent(location.entity(), ignored -> new TreeSet<>())
                     .add(new RuinedLocation(ruinedValue, location.index()));
+            entityToInsertedValuesMap.computeIfAbsent(location.entity(), ignored -> new ArrayList<>()).add(ruinedValue);
+        }
+
+        for (var entry : entityToInsertedValuesMap.entrySet()) {
+            recordingScoreDirector.recordListAssignment(listVariableDescriptor, entry.getKey(), entry.getValue());
```

Comment:
Is the call to `recordingScoreDirector.afterListVariableChanged` no longer necessary?

Reply:
No. At this point, the only thing left to do is to record the assignments, because the CH has already made all the changes. The assignments do need to be recorded though, because we need to be able to undo them later.